Using node pull credentials during builds, image stream imports and pull-through (#136)
openshift/enhancements: enhancements/node-pull-credentials/pull-credentials.md

---
title: node-artifacts-during-build-and-image-stream-import
authors:
  - "@rmarasch"
reviewers:
  - "@dmage"
  - "@bparees"
  - "@adambkaplan"
approvers:
  - "@dmage"
  - "@bparees"
  - "@adambkaplan"
creation-date: 2019-12-02
last-updated: 2019-12-10

> status: provisional

Reviewer: Let's move this to implementable
```suggestion
status: implementable
```

---

# Using node pull credentials during build and image stream import

## Release Signoff Checklist

> - [ ] Enhancement is `implementable`

Reviewer: I'd say this is implementable!
```suggestion
- [x] Enhancement is `implementable`
```

- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

Allow OpenShift users to import and use images from any registry configured
during or after the cluster installation by sharing node's pull credentials

> with `openshift-api` and `builder` pods.

Reviewer 2: @dmage not sure what you mean, how/when would the image-registry be using these creds?  pull-through?

> with `openshift-api` and `builder` pods.

Reviewer 3: @bparees yes, pull-through. My expectation is that if the image can be imported by openshift-apiserver, it should be also accessible via pull-through. If we don't implement this for image-registry, we'll make UX less consistent.

> with `openshift-api` and `builder` pods.

Reviewer 2: yes it's a good catch, the registry should also mount+consume the node creds for pull-through.

> with `openshift-api` and `builder` pods.

Author: Added this as a goal and as a Test Plan scenario. PTAL.

> with `openshift-api` and `builder` pods.

Reviewer: @ricardomaraschini think we need a section for the implementation. Would follow what we would do for openshift-apiserver (operator mounts the hostPath read-only, pull secret is consumed by the image-registry).

> with `openshift-api` and `builder` pods.

Author: Nice catch! I completely forgot about these two items: `readOnly` and a small section to describe how we see pull-through working. Have added both to the proposal.

## Motivation

To have image stream and builds working closely to node avoids complexity and
redundancy as each component won't need to mount different Secrets to have

> access to information that is already present on node's filesystem.

Reviewer: nit
```suggestion
access to information that is already present on the node's filesystem.
```

In order to provide a seamless developer experience OpenShift should allow
users to import or to use during builds any registry credentials provided
during or after the installation, among the registries we would like to allow

> users to interact with is `registry.redhat.io`. 

Reviewer: Cleaning this up a bit:
```
OpenShift should seamlessly use the cluster-wide pull secret provided during installation
in builds and imagestream imports. This is particularly important for images pulled from
`registry.redhat.io`, which requires a pull secret.
```

Author: done.

> Nowadays pull credentials provided during cluster install are available on

Reviewer: nit
```suggestion
Today pull secrets provided during cluster install are available on the
```

node's filesystem. If user attempts to import an image stream from these

> locations OpenShift fails as neither `openshift-api` nor `builder` use the

Reviewer: nit
```suggestion
locations, OpenShift fails as neither `openshift-api` nor `builder` use the
```

credentials provided during the installation.

### Goals

- Allow users to import image streams from images hosted in any registry

>   provided during or after cluster installation, including `registry.redhat.io`.

Reviewer 3: Allow users to use container images from any registry provided during or after cluster installation

Author: fixed.

- Allow users to run `builds` based on images hosted in any registry provided
  during or after cluster installation without providing any extra credentials.

### Non-Goals

- Change the way users manage cluster pull credentials.
  * Additional credentials may still be provided through secrets on namespace.
- Use Node mirroring config.
  * To use config files files such as `/etc/containers/registries.conf` is not
  part of this proposal.
- Use Node CAs.
  * This should be done in the future through another enhancement proposal as

>   we currently have different sources of truth when it comes to CAs(different

Reviewer: nit
```suggestion
  we currently have different sources of truth when it comes to CAs (different
```

  methods of providing them to different parts of the codebase).

## Proposal

### User Stories

As an OpenShift user
I want to be able to import imagestreams from any image registry that the
cluster nodes can pull images from

> So that I can use them without manually creating any extra credentials

Reviewer 3: nit, markdown will not render this on a new line
```suggestion
so that I can use them without manually creating any extra credentials
```

As an OpenShift user
I want to be able to use, during builds, images from any image registry that
the cluster nodes can pull images from

> So that I can use these images as base for my own images

Reviewer 3: ```suggestion
so that I can use these images as base for my own images
```

### Implementation Details/Notes/Constraints

This is the node's filesystem path we are taking into account:

> - `/var/lib/kubelet/config.json` contains node's pull credentials.

Reviewer: nit
```suggestion
- `/var/lib/kubelet/config.json` contains the node's pull secret.
```

#### Image Stream Import

> - Mount the location as `hostPath` during `openshift-apiserver` deploy under

Reviewer: nit
```suggestion
- Mount the pull secret as a `hostPath` in the `openshift-apiserver` deployment under
```

  `/node/var/lib/kubelet/config.json`.

> ```

Reviewer: markdown nit
```suggestion
```yaml
```

volumeMounts:
- mountPath: /node/var/lib/kubelet/config.json
  name: node-pull-credentials
volumes:
- name: node-pull-credentials
  hostPath:
    path: /var/lib/kubelet/config.json
    type: File
```

- When user requires to import an image stream, `openshift-apiserver` would
  merge all pull credentials found in `/node/var/lib/kubelet/config.json` with

>   other credentials that may exist on the namespace.

Reviewer: Cleaning up a bit:

```
- When a user imports an imagestream, `openshift-apiserver` will
  merge all pull credentials found in `/node/var/lib/kubelet/config.json` with
  other credentials that may exist in the namespace.
```

Author: done.

#### Builds

- `controller-manager` would mount pull credentials inside `build` pod as 
  `hostPath`(similar to what is done for Image Stream Import).
- Builder image parses node's pull credentials and uses them during build,

>   merging with other pull credentials that may exist on the same namespace.

Reviewer: Adding a clarifying note about pull secrets in the `BuildConfig`:

```
- Builder image parses node's pull credentials and uses them during build,
  merging with other pull credentials that are linked to the `builder` service account.
  If the `BuildConfig` specifies a pull secret, we will continue the current behavior of
  using the provided pull secret as an override.
```

Author: done.

> ### Risks and Mitigations

Reviewer 3: Another risk: these credentials may leak through ImageStreamSecret API

Author: Could you clarify?

Author: After our discussions I added an extra item to our "Risks and Mitigations" session. PTAL [here](https://github.com/openshift/enhancements/pull/136/files#diff-a3521cf1ef50ae6c95913acbdca6106dR162-R174).

#### Pull credentials inside build pod may be visible to the user

> Cluster wise credentials mounted inside the builder pod may be a security risk

Reviewer 2: ```suggestion
Cluster wide credentials mounted inside the builder pod may be a security risk
```

as the user may be able to shell into the pod and copy them. Another possible
risk is that the user may copy the credentials into a resulting image.

Mitigations:

As far as I verified(and this needs to be once more tested) it is impossible
to spawn a shell inside the builder pods. I also tried to copy the credentials
from builder's filesystem into a resulting image but it failed.

#### Pushing to registries using node's credentials

If pull credentials are used during builds we may allow users to push images to
any registry configured on the node. This is not an ideal scenario as cluster
administrators may not to be aware or wanting this.

Mitigations

Pull credentials must be used **only** for pulling images. On pushing we must

> use only default credentials'.

Reviewer 2: ```suggestion
use only user provided(not node) credentials'.
```

#### Absence of mounted path on node filesystem

If the path we are trying to mount through `hostPath` directive does not exist
on the node where the pod is running the pod won't come up.

Mitigations

> Mounted path always exist on master nodes(_still awaiting for node's team check

Reviewer 2: we don't just run build pods on master nodes, we run them on worker nodes (in fact i doubt we run them on master nodes at all currently..i do not think we tolerate that taint, but that taint could be removed in the future)

Author: fixed.

## Motivation

To have image stream, image registry and builds working closely to node avoids
complexity and redundancy as each component won't need to mount different
Secrets to have access to information that is already present on the node's
filesystem.

OpenShift should seamlessly use the cluster-wide pull secret provided during
installation in builds, imagestream imports and pull-through operations. This
is particularly important for images pulled from `registry.redhat.io`, which
requires a pull secret.

Today pull secrets provided during cluster install are available on the
node's filesystem. If user attempts to import an image stream or pull-through
from these locations, OpenShift fails as none of `openshift-api`, `builder` or
`image-registry` use the credentials provided during the installation.

### Goals

- Allow users to use container images from any registry provided during or
  after cluster installation.
- Allow users to run `builds` based on images hosted in any registry provided
  during or after cluster installation without providing any extra credentials.
- Allow `image-registry` to execute pull-through using the node's pull

>   credentials.

Reviewer 4: What I'm missing here is an explicit opt-in for that functionality, not every cluster-wide secret can and should be exposed. I see the requirement, but the risks are much higher, so this should not be a default, rather an opt-in.

Reviewer: @soltysh this enhancement is narrowly focused on the `pull-secret` added to the `openshift-config` namespace at install time. This pull secret is implicitly shared cluster-wide on all nodes - I am not aware of any opt-out other than deleting this secret.

I agree that a general "share any secret across a cluster" feature should absolutely be opt-in. DevEx has discussed this a bit as a future enhancement proposal [1]

[1] https://issues.redhat.com/browse/DEVEXP-426

  part of this proposal.
- Use Node CAs.
  * This should be done in the future through another enhancement proposal as
  we currently have different sources of truth when it comes to CAs (different
  methods of providing them to different parts of the codebase).

## Proposal

### User Stories

As an OpenShift user
I want to be able to import imagestreams from any image registry that the
cluster nodes can pull images from
so that I can use them without manually creating any extra credentials.

As an OpenShift user
I want to be able to use, during builds, images from any image registry that
the cluster nodes can pull images from
so that I can use these images as base for my own images.

As an OpenShift user
I want to be able to pull-through images from any registry that the cluster
nodes can pull images from
so that I can use image registry cache to speed-up my builds.

### Implementation Details/Notes/Constraints

This is the node's filesystem path we are taking into account:

- `/var/lib/kubelet/config.json` contains the node's pull secret.

#### Image Stream Import

- Mount the pull secret as `readOnly`, `hostPathreadOnly` in

>   `openshift-apiserver` deployment under `/node/var/lib/kubelet/config.json`.

Author: @dmage we need to come up with an agreement when it comes to the place where we are going to mount this. What is your idea?

Reviewer 3: my idea is to use the same path inside the pod as on the node, but I'd like to hear ideas from someone who can approve PRs for openshift-apiserver.

Reviewer 3: and I'd say it should be configurable, i.e. the operator should mount `/var/lib/kubelet/config.json` and pass this name into the configuration file.

Reviewer 5: @csrwng we would also need to track mounting this into the hypershift-toolkit.

Reviewer 6: Ack

Author: Following @dmage suggestion I have updated the proposal, now we mount pull secrets on `/var/lib/kubelet/config.json`

  merging with other pull credentials that are linked to the `builder` service
  account. If the `BuildConfig` specifies a pull secret, we will continue the
  current behavior of using the provided pull secret as an override.

#### Registry pull-through

- As done for Image Stream Import, mount pull credentials inside the image
  registry pod.
- Pull credentials will then be consumed by the registry.

### Risks and Mitigations

#### Pull credentials for registry may exist on namespace and on node

User may have created a secret containing credentials for a registry that node
already has credentials to.

Mitigations:

Always prioritize namespace secrets over node's credentials. If a credential
exists on the namespace, do not use node's credentials.

#### Pull credentials inside build pod may be visible to the user

Cluster wide credentials mounted inside the builder pod may be a security risk
as the user may be able to shell into the pod and copy them. Another possible
risk is that the user may copy the credentials into a resulting image.

Mitigations:

As far as I verified(and this needs to be once more tested) it is impossible

> to spawn a shell inside the builder pods. I also tried to copy the credentials

Reviewer 4: Ephemeral pods will allow sharing process namespace and thus provide you with access to it. That's a major problem.

Reviewer 4: See https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/

Author: This issue also applies to "openshift-apiserver" during image stream import or is something only for builds?

Reviewer 2: @ricardomaraschini what @soltysh is saying is that even though we block oc rsh into a build-pod, someone (may) be able to create an ephemeral container in the build-pod's process namespace and use that mechanism to get into it instead.

@deads2k are there security mechanisms that would prevent someone from abusing this?  much like we don't allow you to exec/rsh into a pod you can't "create" yourself (so normal users can't exec/rsh into a privileged container, as i understand it), does the ephemeral container api have similar restrictions?

Reviewer 5: @bparees there is no limitation at this time to my knowledge.  its not clear to me that we could use rbac appropriately in a namespace to block the subresource to create the ephemeral container as build pods would be in the same namespace as other pods we would want to allow ephemeral access.

Reviewer 2: that subresource effectively updates the pod resource.  If i can't create privileged pods, i would not expect to be able to edit prvileged pods and i'd expect that restriction to extend to subresources that modify the pod.  No?

Reviewer 2: spoke with @derekwaynecarr offline.  SCC will need to handle this (ensure that a user who can't create a particular pod also cannot invoke this subresource which edits the pod).

between @openshift/sig-developer-experience and @openshift/sig-master we need to ensure this is covered by the time the ephemeral container api goes beta (at which point it will be on by default).

Author: I have enabled EphemeralContainers on kube-apiserver and simulated the feature, as `kubeadmin` I have been able to patch the pod normally as expected but as a regular user I  received the following message back:

```
$ oc replace --raw /api/v1/namespaces/build/pods/img-1-build/ephemeralcontainers -f ./ec.json 
Error from server (Forbidden): pods "img-1-build" is forbidden: User "user" cannot update resource "pods/ephemeralcontainers" in API group "" in the namespace "build"
$
```

ec.yaml
```
$ cat ec.json 
{
    "apiVersion": "v1",
    "kind": "EphemeralContainers",
    "metadata": {
            "name": "img-1-build"
    },
    "ephemeralContainers": [{
        "command": [
            "sh"
        ],
        "image": "busybox",
        "imagePullPolicy": "IfNotPresent",
        "name": "debugger",
        "stdin": true,
        "tty": true,
        "terminationMessagePolicy": "File"
    }]
}
```

exists on the namespace, do not use node's credentials.

#### Pull credentials inside build pod may be visible to the user

Cluster wide credentials mounted inside the builder pod may be a security risk
as the user may be able to shell into the pod and copy them. Another possible
risk is that the user may copy the credentials into a resulting image.

Mitigations:

As far as I verified(and this needs to be once more tested) it is impossible
to spawn a shell inside the builder pods. I also tried to copy the credentials
from builder's filesystem into a resulting image but it failed.

#### Pushing to registries using node's credentials

If pull credentials are used during builds we may allow users to push images to
any registry configured on the node. This is not an ideal scenario as cluster
administrators may not to be aware or wanting this.

Mitigations

Pull credentials must be used **only** for pulling images. On pushing we must
use only user provided(not node) credentials'.

#### Absence of mounted path on node filesystem

If the path we are trying to mount through `hostPath` directive does not exist
on the node where the pod is running the pod won't come up.

Mitigations

> Mounted path always exist on worker nodes.

Reviewer 7: what about master nodes?

Author: Interesting. The confirmation we have is that this path does not exist [only on bootstrap nodes](https://github.com/openshift/enhancements/pull/136#discussion_r368620589). @adambkaplan could we tag someone from node's team to get a final answer on this?

Reviewer: @cgwalters can you please confirm that the cluster pull secret always exists at `/var/lib/kubelet/config.json`

Reviewer 8: I don't want to make that permanent "ABI" though; see also https://github.com/openshift/machine-config-operator/issues/1190 which proposes ensuring all MCO configuration (including pull secret) lives in `/etc` so it can be made transactional - (but we'd probably have a backcompat symlink for a while at least).

I think probably the best bet is for whatever wants it to parse the kubelet config or so perhaps?

Or, I guess we could make `/etc/kubernetes/pull-secret.json -> /var/lib/kubelet/config.json` today and people could start using the new name.  (Or probably better introduce the that new `/etc` name upstream)

Author: > I don't want to make that permanent "ABI" though; see also [openshift/machine-config-operator#1190](https://github.com/openshift/machine-config-operator/issues/1190) which proposes ensuring all MCO configuration (including pull secret) lives in `/etc` so it can be made transactional - (but we'd probably have a backcompat symlink for a while at least).

It is clear that we must make the file path easily configurable.
 
> I think probably the best bet is for whatever wants it to parse the kubelet config or so perhaps?

@cgwalters Do you mean to parse `/etc/kubernetes/kubelet.conf` to know from where to read pull secrets? I could not find any directive on it pointing to `/var/lib/kubelet/config.json`, pardon if I misunderstood you, could you clarify?

> Or, I guess we could make `/etc/kubernetes/pull-secret.json -> /var/lib/kubelet/config.json` today and people could start using the new name. (Or probably better introduce the that new `/etc` name upstream)

I guess this would take some time to get done and we already have some PRs in place to implement this enhancement. I am inclined to stick with `/var/lib/kubelet/config.json`, making it easily configurable so we can easily migrate in the future.

Reviewer 3: @cgwalters are you going to change root-dir for kubelet?

  * Additional credentials may still be provided through secrets on namespace.
- Use Node mirroring config.
  * To use config files files such as `/etc/containers/registries.conf` is not
  part of this proposal.
- Use Node CAs.
  * This should be done in the future through another enhancement proposal as
  we currently have different sources of truth when it comes to CAs (different
  methods of providing them to different parts of the codebase).

## Proposal

### User Stories

As an OpenShift user
I want to be able to import imagestreams from any image registry that the
cluster nodes can pull images from
so that I can use them without manually creating any extra credentials.

As an OpenShift user
I want to be able to use, during builds, images from any image registry that
the cluster nodes can pull images from
so that I can use these images as base for my own images.

As an OpenShift user
I want to be able to pull-through images from any registry that the cluster
nodes can pull images from
so that I can use image registry cache to speed-up my builds.

### Implementation Details/Notes/Constraints

This is the node's filesystem path we are taking into account:

> - `/var/lib/kubelet/config.json` contains the node's pull secret.

Reviewer 7: can this file change? E.g. a user can ssh into the machine and append secrets there.

Reviewer 7: This has influence on https://github.com/openshift/cluster-openshift-apiserver-operator/pull/284, i.e. we need to reread that file. Is that the case?

Reviewer 2: it can be changed via the MCO which causes node reboots so rereading is moot.

if the MCO is ever updated to allow node filesystem changes w/o rebooting the nodes, we'd have to setup a watch on the mounted file in our pod.

I believe these steps are what lead to the /var/lib/kubelet/config.json being updated:
https://docs.openshift.com/container-platform/4.3/openshift_images/managing-images/using-image-pull-secrets.html#images-update-global-pull-secret_using-image-pull-secrets

Author: Pull secrets are read in a "per image stream import request", we don't keep it on memory.

Cluster wide credentials mounted inside the builder pod may be a security risk
as the user may be able to shell into the pod and copy them. Another possible
risk is that the user may copy the credentials into a resulting image.

Mitigations:

As far as I verified(and this needs to be once more tested) it is impossible
to spawn a shell inside the builder pods. I also tried to copy the credentials
from builder's filesystem into a resulting image but it failed.

#### Pushing to registries using node's credentials

If pull credentials are used during builds we may allow users to push images to
any registry configured on the node. This is not an ideal scenario as cluster
administrators may not to be aware or wanting this.

Mitigations

Pull credentials must be used **only** for pulling images. On pushing we must
use only user provided(not node) credentials'.

#### Absence of mounted path on node filesystem

If the path we are trying to mount through `hostPath` directive does not exist
on the node where the pod is running the pod won't come up.

Mitigations

Mounted path always exist on worker nodes(_still awaiting for node's team check

> on this regard_).

Reviewer 2: are we still waiting on this confirmation?

Author: It would be good to have this confirmation, yes.

Author: @umohnani8 @mrunalp we are waiting for an OK on this path here, could you help us out?

Reviewer 3: as per Urvashi, it should always be there (expect maybe for bootstrap nodes).

AFAIK we don't need them on bootstrap nodes, so we are fine.

Author: OK, resolving this one.
